pmlc/rt/boundscheck.cc

```diff
@@ -0,0 +1,15 @@
+// Copyright 2020 Intel Corporation
+
+#include <cstdint>
+#include <stdexcept>
+
+// This function is called from the BoundsCheckPass.
+// index - index for a dimension that is accessed by load/store operation
+// range - upper bound of the range for this dimension (it is always 0 -
+// range)
+extern "C" void plaidml_rt_bounds_check(intptr_t index, int64_t range) {
+  int64_t accessIndex = static_cast<int64_t>(index);
+  if (accessIndex < 0 || accessIndex >= range)
+    throw std::runtime_error(
+        "Out of bounds index for mlir::LoadOp or mlir::StoreOp");
```

Comment:
I think we should have a test that purposefully goes out of bounds to ensure that this is all working as we expect. We should be able to do that by hand writing the MLIR code to go out of bounds.

Reply:
This kind of test is certainly easy to do, but we can't run anything yet - need to split the conversion to Stdx and LLVM dialects and insert this pass there. I certainly am planing to add a test like this. Once this is integrated maybe? I just want to add the method in for now, so I can work on splitting the conversion and adding this pass to the pipeline.

Reply:
Integration should involve the test, why add the integration without the test? In general, I think we should practice TDD (Test-Driven Development), where we first add a test that is broken and expected to fail because we haven't written any code. Then add the code to get the test to pass, and then finally commit and get it reviewed.

Reply:
FYI, the final pipeline for x86 does not need to be complete, that can be integrated later. What's great about MLIR is that you can piece together parts of passes using `pmlc-opt` and `pmlc-jit` command line utilities and then put this into a `.mlir` test. See the `xsmm_call.mlir` test for an example of how to put together a sequence of passes along with running it in the JIT and then verifying that we get the result we want to have.

Reply:
The RUN line would look like:

```
// RUN: pmlc-opt -stdx-check-bounds -convert-std-to-llvm %s | pmlc-jit -entry-point-result=void | FileCheck %s
```

And then you'll need to add a main entry:

```
func @main() {
  %i = constant 0 : index
  %j = constant 10 : index // out of bounds
  %buf = alloc() : memref<20x10xf32>
  call @simpleLoad(%buf, %i, %j) -> (f32) : (memref<20x10xf32>, index, index) -> (f32)
  dealloc %buf : memref<20x10xf32>
}
```

Then we need to figure out how to check for the exception being thrown.

Reply:
Added a test and support in plmc-jit for allowing to specify expected runtime exception.